### PR TITLE
Bug 1757708: CVE-2019-11253 billion laughs

### PR DESCRIPTION
**Pull request by jottofar**
Process followed:
1. Modified and committed glide.yaml

2. Modified glide.yaml per 3.11 to contain https://github.com/openshift/gssapi.git version release-2.6.3. Pinned github.com/containers/image, github.com/emicklei/go-restful-swagger12, github.com/onsi/ginkgo, github.com/opencontainers/runc, github.com/openshift/api, github.com/openshift/client-go, github.com/vmware/photon-controller-go-sdk and committed.

3. 'make update-deps' and committed bump(*)
 
4. UPSTREAM: 73713: Importing the latest json patch and set the accumulated copy size limit commit

5. UPSTREAM: 73805: Adding a limit on the size of request body commit

6. Cherry-picked 5cb96b8 as "UPSTREAM: 83261..." commit

7. UPSTREAM: <carry>: fixup tests after 83261 commit

### CODE REVIEW

**Comment by jottofar** [comment]
/review @deads2k 
Doesn't have the required unit test case changes yet but want to get it out here to be sure it builds and if you want to start reviewing.

**Comment by deads2k** [comment]
@eparis is the canonical source here or in OSE?

**Comment by deads2k** [comment]
> /review @deads2k
> Doesn't have the required unit test case changes yet but want to get it out here to be sure it builds and if you want to start reviewing.

wow, CI is borked :(

**Comment by jottofar** [comment]
the 3.10 merge this depends on is not complete yet so it failed

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -77,6 +77,10 @@ type crdHandler struct {
 	delegate          http.Handler
 	restOptionsGetter generic.RESTOptionsGetter
 	admission         admission.Interface
+
+	// The limit on the request size that would be accepted and decoded in a write request
+	// 0 means no limit.
+	maxRequestBodyBytes int64
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: This is in the wrong commit. What is this for?

File: vendor/github.com/containers/image/openshift/openshift.go [hunk 2]
```diff
@@ -136,7 +136,7 @@ func (c *openshiftClient) doRequest(ctx context.Context, method, path string, re
 // getImage loads the specified image object.
 func (c *openshiftClient) getImage(ctx context.Context, imageStreamImageName string) (*image, error) {
 	// FIXME: validate components per validation.IsValidPathSegmentName?
-	path := fmt.Sprintf("/oapi/v1/namespaces/%s/imagestreamimages/%s@%s", c.ref.namespace, c.ref.stream, imageStreamImageName)
+	path := fmt.Sprintf("/apis/image.openshift.io/v1/namespaces/%s/imagestreamimages/%s@%s", c.ref.namespace, c.ref.stream, imageStreamImageName)
```
Comment (deads2k) [inline vendor/github.com/containers/image/openshift/openshift.go]: this looks wrong

File: vendor/github.com/docker/docker/client/hijack.go [hunk 3]
```diff
@@ -206,10 +206,3 @@ type hijackedConn struct {
 func (c *hijackedConn) Read(b []byte) (int, error) {
 	return c.r.Read(b)
 }
-
-func (c *hijackedConn) CloseWrite() error {
```
Comment (deads2k) [inline vendor/github.com/docker/docker/client/hijack.go]: this looks off

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go [hunk 4]
```diff
@@ -270,13 +271,19 @@ func (m *Manager) Apply(pid int) error {
...
-		// systemd converts CPUQuotaPerSecUSec (microseconds per CPU second) to CPUQuota
-		// (integer percentage of CPU) internally.  This means that if a fractional percent of
-		// CPU is indicated by Resources.CpuQuota, we need to round up to the nearest
-		// 10ms (1% of a second) such that child cgroups can set the cpu.cfs_quota_us they expect.
-		if cpuQuotaPerSecUSec%10000 != 0 {
-			cpuQuotaPerSecUSec = ((cpuQuotaPerSecUSec / 10000) + 1) * 10000
+		// corresponds to USEC_INFINITY in systemd
+		// if USEC_INFINITY is provided, CPUQuota is left unbound by systemd
+		// always setting a property value ensures we can apply a quota and remove it later
+		cpuQuotaPerSecUSec := uint64(math.MaxUint64)
```
Comment (deads2k) [inline vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: this looks suspicious.

File: vendor/k8s.io/kubernetes/pkg/kubelet/dockershim/docker_image.go [hunk 5]
```diff
@@ -96,23 +96,30 @@ func (ds *dockerService) RemoveImage(image *runtimeapi.ImageSpec) error {
...
 	// dockerclient.InspectImageByID doesn't work with digest and repoTags,
 	// it is safe to continue removing it since there is another check below.
 	if err != nil && !libdocker.IsImageNotFoundError(err) {
 		return err
 	}
 
-	_, err = ds.client.RemoveImage(image.Image, dockertypes.ImageRemoveOptions{PruneChildren: true})
-	if err != nil && !libdocker.IsImageNotFoundError(err) {
-		return err
+	if imageInspect == nil {
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/kubelet/dockershim/docker_image.go]: I don't understand why this changed.

**Comment by jottofar** [comment]
/review @sttts @eparis can I get one or both to review please. @deads2k had questions about my bump.

File: vendor/github.com/containers/image/openshift/openshift.go (hunk 2 above)
Comment (jottofar) [reply vendor/github.com/containers/image/openshift/openshift.go]: I don't know why the bump picked up the files in github.com or in k8s.io unlike the others I did for 3.10, 3.11, etc.

File: vendor/github.com/docker/docker/client/hijack.go (hunk 3 above)
Comment (jottofar) [reply vendor/github.com/docker/docker/client/hijack.go]: Again, agree, don't know why the bump picked up the files in github.com or in k8s.io

File: vendor/github.com/containers/image/openshift/openshift.go (hunk 2 above)
Comment (sttts) [reply vendor/github.com/containers/image/openshift/openshift.go]: @deads2k merged by you into the 3.9 branch https://github.com/openshift/containers-image/commit/e632a5797c121e35233eedc7cae477844e3d760a. The problem is that we use the 3.9 branch even today.

**Comment by jottofar** [comment]
@deads2k @sttts 
-redid the bump to pin some stuff so it didn't pick up invalid changes
- also separated out some changes into a spearate commit hoping iot would fix the checkcommiter issue but did not - STUCK ON commitchecker failure

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/2.edited [hunk 6]
```diff
@@ -1,7 +1,7 @@
 # Please edit the 'last-applied-configuration' annotations below.
 # Lines beginning with a '#' will be ignored, and an empty file will abort the edit.
 #
-# The edited file had a syntax error: error converting YAML to JSON: yaml: line 12: could not find expected ':'
+# The edited file had a syntax error: error converting YAML to JSON: yaml: line 13: could not find expected ':'
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/2.edited]: these files are UPSTREAM changed. Why are they there in the first place?

Comment (jottofar) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/2.edited]: In the last two PR's, 3.10 and 3.11, @deads2k had me keep the required unit test changes separate so they could be easily checked. I just continued that pattern to be consistent but I can see the reasoning of just combining now if you prefer.

**Comment by jottofar** [comment]
@deads2k @sttts based on the commitchecker errors I tried:

- ensure no other, non upstream files are included in an UPSTREAM commit
- divided each, what I think are,  different upstream repos - json-iterator, modern-go and kubernetes, into their own separate commits. I may not have divided all them.
- changed all the above commit messages to contain 'UPSTREAM: 83261:'. Not sure this is correct but that's the last thing I tried.

**Comment by sttts** [comment]
Why not do what the commitchecker actually proposes in https://travis-ci.com/openshift/origin/builds/133081030 ?

- The following UPSTREAM commits modify more than one repo in their changelist: [31875775c6] UPSTREAM: 83261: modern-go dependency

- The following UPSTREAM commits modify Godeps repos other than the repo the commit declares:[31875775c6] UPSTREAM: 83261: modern-go dependency

**Comment by jottofar** [comment]
@sttts  answer was in front of me whole time, I had noticed in code that commitchecker skips checking commits that begin with 'bump(' and also just noticed that prior updates to modern-go and json-iterator had commit message bump(*).

File: glide.yaml [hunk 7]
```diff
@@ -40,9 +40,9 @@ import:
 
 # openshift second
 - package: github.com/openshift/api
-  version: release-3.9
+  version: 0d921e363e951d89f583292c60d013c318df64dc
```
Comment (sttts) [inline glide.yaml]: these glide.yaml changes should not be in a bump commit, but separate.

File: glide.yaml [hunk 8]
```diff
@@ -353,6 +353,8 @@ import:
   version: 72f9bd7c4e0c2a40055ab3d0f09654f730cce982
 - package: github.com/juju/ratelimit
   version: 5b9ff866471762aa2ab2dced63c9fb6f53921342
+- package: github.com/json-iterator/go
+  version: f2b4162afba35581b6d4a50d3b8f34e33c144682
```
Comment (sttts) [inline glide.yaml]: same as before, separate these out into another commit.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/2.edited (hunk 6 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/testdata/edit/testcase-apply-edit-last-applied-syntax-error/2.edited]: yes, extra commit please.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 9]
```diff
@@ -111,6 +116,7 @@ func NewCustomResourceDefinitionHandler(
 		delegate:                delegate,
 		restOptionsGetter:       restOptionsGetter,
 		admission:               admission,
+		maxRequestBodyBytes:     maxRequestBodyBytes,
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: missing the use of this further down in the request scope https://github.com/openshift/origin/pull/23987/commits/5f027c5c20e3a8fdc3cc7acf92a46bd47bc4cdea#diff-b4feb82ec380748d1f9374cbf25bfc7bL372

Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Compare https://github.com/kubernetes/kubernetes/pull/83436/commits/a4e734ac9fea9affdffee9ac845c15e1abacfab3#diff-2f720a7e61a4ddfcfe9a9e353be8784bR594 in upstream.

Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: To address this I needed to essentially cherry pick " UPSTREAM: 73805: Adding a limit on the size of request body the apiserver will decode for write operations" which was added in 3.11 and not yet back-ported.

File: glide.yaml (hunk 8 above)
Comment (jottofar) [reply glide.yaml]: In order to quickly address I made the minimum change here which was to carve out change to glide.yaml from bump into it's own commit "pinned depedndencies"

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 9 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Why was it left out in the first place?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 10]
```diff
@@ -541,6 +541,8 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		Kind:        fqKindToRegister,
 
 		MetaGroupVersion: metav1.SchemeGroupVersion,
+
+		MaxRequestBodyBytes: a.group.MaxRequestBodyBytes,
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: is this now a mix of two upstream commits? This line came from the other commit, didn't it?

Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go]: oh, this is not customresource_handler, nevermind

File: vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/options/options_test.go [hunk 11]
```diff
@@ -109,6 +109,8 @@ func TestAddFlags(t *testing.T) {
 			MaxMutatingRequestsInFlight: 200,
 			RequestTimeout:              time.Duration(2) * time.Minute,
 			MinRequestTimeout:           1800,
+			JSONPatchMaxCopyBytes:       int64(3 * 1024 * 1024),
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/cmd/kube-apiserver/app/options/options_test.go]: this is not the upstream commit, but adapted to the different order. This makes it hard to follow and compare. Better pick in the same order as upstream.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go [hunk 12]
```diff
@@ -52,6 +60,8 @@ func NewServerRunOptions() *ServerRunOptions {
 		MaxMutatingRequestsInFlight: defaults.MaxMutatingRequestsInFlight,
 		RequestTimeout:              defaults.RequestTimeout,
 		MinRequestTimeout:           defaults.MinRequestTimeout,
+		JSONPatchMaxCopyBytes:       defaults.JSONPatchMaxCopyBytes,
+		MaxRequestBodyBytes:         defaults.MaxRequestBodyBytes,
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: upstream has only one value in this commit. Why another one? Was this something forgotten in the first commit?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 9 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: added below

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 13]
```diff
@@ -493,19 +496,6 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
...
-		}
-		existing := atomic.LoadInt64(&jsonpatch.AccumulatedCopySizeLimit)
-		if existing > 0 && existing < c.JSONPatchMaxCopyBytes {
-			break
-		}
-		if atomic.CompareAndSwapInt64(&jsonpatch.AccumulatedCopySizeLimit, existing, c.JSONPatchMaxCopyBytes) {
-			break
-		}
-	}
-
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go]: this shouldn't be here. It was added in the first UPSTREAM commit and is removed here again.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 14]
```diff
@@ -68,32 +71,75 @@ var _ recognizer.RecognizingDecoder = &Serializer{}
...
+			i64, err := strconv.ParseInt(string(number), 10, 64)
+			if err == nil {
+				*(*interface{})(ptr) = i64
+				return
+			}
+			f64, err := strconv.ParseFloat(string(number), 64)
+			if err == nil {
+				*(*interface{})(ptr) = f64
+				return
+			}
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: compare upstream (1.13 PR) https://github.com/kubernetes/kubernetes/pull/83436/commits/a4e734ac9fea9affdffee9ac845c15e1abacfab3

Upstream has a DecodeNumber error here.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 15]
```diff
@@ -68,32 +71,75 @@ var _ recognizer.RecognizingDecoder = &Serializer{}
...
+				}
+			}
+
+			*(*interface{})(ptr) = iter.Read()
+
+			// restore current depth
+			iter.Attachment = originalAttachment
+		}
+	}
+	jsoniter.RegisterTypeDecoderFunc("interface {}", decodeNumberAsInt64IfPossible)
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: this line I don't see in upstream. https://github.com/kubernetes/kubernetes/pull/83436/commits/a4e734ac9fea9affdffee9ac845c15e1abacfab3#diff-f216f544515d2fd05d66d92c5f95a248R123

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 16]
```diff
@@ -68,32 +71,75 @@ var _ recognizer.RecognizingDecoder = &Serializer{}
...
+func CaseSensitiveJsonIterator() jsoniter.API {
+	return jsoniter.Config{
+		EscapeHTML:             true,
+		SortMapKeys:            true,
+		ValidateJsonRawMessage: true,
+		CaseSensitive:          true,
+	}.Froze()
+}
+
+var caseSensitiveJsonIterator = CaseSensitiveJsonIterator()
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: where is this used?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go [hunk 17]
```diff
@@ -68,32 +71,75 @@ var _ recognizer.RecognizingDecoder = &Serializer{}
...
+
+// gvkWithDefaults returns group kind and version defaulting from provided default
+func gvkWithDefaults(actual, defaultGVK schema.GroupVersionKind) schema.GroupVersionKind {
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: where is this used?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 13 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go]: Add this back in?

for {
		if c.JSONPatchMaxCopyBytes <= 0 {
			break
		}
		existing := atomic.LoadInt64(&jsonpatch.AccumulatedCopySizeLimit)
		if existing > 0 && existing < c.JSONPatchMaxCopyBytes {
			break
		}
		if atomic.CompareAndSwapInt64(&jsonpatch.AccumulatedCopySizeLimit, existing, c.JSONPatchMaxCopyBytes) {
			break
		}
}

Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go]: yes

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go (hunk 14 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: I tried to put in minimal change that would meet the intent of the upstream change, i.e. not perturb the original code if it wasn't involved in the new logic. I thought the core logic for upstream change is by adding 'default:' case and checking increment depth. Why mess with case NumberValue?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go (hunk 15 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: It comes from 3.11, then I back ported into 3.10 and finally into 3.9. It appears to be crucial to getting this logic to run at all in 3.10 and 3.9 where the actual upstream change then comes into play.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go (hunk 16 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: It is not so will remove, inadvertently pick from 3.10 in which it is also not used. Original source was inadvertently picked from 3.11.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go (hunk 17 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go]: is not, will remove

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 13 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go]: This needs a later version of json-patch as current version does not include AccumulatedCopySizeLimit. What is the correct way to do that? Use the sha from the 3.10 glide.lock?

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go (hunk 12 above)
Comment (jottofar) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go]: TargetRAMB was already there in 3.9. The change is to align it's declaration with MaxRequestBodyBytes.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 13 above)
Comment (sttts) [reply vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go]: yes, maybe the 3.10 version.

**Comment by sttts** [comment]
e2e-gcp is also broken in master: https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/24038/test_pull_request_origin_extended_conformance_gce_39/1239/

/skip gcp

**Pull request merged**